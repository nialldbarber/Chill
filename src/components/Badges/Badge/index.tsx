export {default as Badges} from './Badge';
