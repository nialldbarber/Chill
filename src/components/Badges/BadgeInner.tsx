import React from 'react';
import {StyleSheet, Text, TouchableOpacity} from 'react-native';
import {useDispatch} from 'react-redux';
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  withSpring,
} from 'react-native-reanimated';
import {useTheme} from '@react-navigation/native';
import {
  widthPercentageToDP as wp,
  heightPercentageToDP as hp,
} from 'react-native-responsive-screen';
import {feelings} from '~/constants/exercises';
import {filterBySelectedBadge} from '../../store/slices/exercises';

type BadgeInnerProps = {
  i: number;
  item: string;
  press: () => void;
};

export default function BadgeInner({i, item, press}: BadgeInnerProps) {
  const {colors} = useTheme();

  const styles = StyleSheet.create({
    badgeContainer: {
      marginTop: hp('2%'),
      marginBottom: hp('4%'),
    },
    badge: {
      backgroundColor: colors.primaryFaded,
      paddingHorizontal: 12,
      paddingVertical: 6,
      borderRadius: 20,
      marginRight: 10,
      borderColor: colors.primary,
      borderWidth: 1,
    },
    badgeText: {
      color: colors.primary,
      fontWeight: '800',
      fontSize: wp('4%'),
    },
  });

  const dispatch = useDispatch();

  const scale = useSharedValue(1);
  const scaleStyles = useAnimatedStyle(() => ({
    transform: [{scale: scale.value}],
  }));

  function handleFilter(): void {
    setTimeout(() => {
      dispatch(filterBySelectedBadge(feelings[i]));
    }, 200);
  }

  return (
    <TouchableOpacity
      key={i}
      activeOpacity={1}
      onPress={() => {
        handleFilter();
        press();
      }}
      onPressIn={() => (scale.value = withSpring(1.05))}
      onPressOut={() => (scale.value = withSpring(1))}
      style={styles.badgeContainer}
    >
      <Animated.View
        style={[
          styles.badge,
          {marginLeft: i === 0 ? wp('5%') : 0},
          scaleStyles,
        ]}
      >
        <Text style={styles.badgeText}>{item}</Text>
      </Animated.View>
    </TouchableOpacity>
  );
}
