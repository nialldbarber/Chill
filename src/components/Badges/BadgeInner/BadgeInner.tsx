import React from 'react';

import {StyleSheet, Text} from 'react-native';
import Animated, {
  useAnimatedStyle,
  useSharedValue,
  withSpring,
} from 'react-native-reanimated';
import {
  heightPercentageToDP as hp,
  widthPercentageToDP as wp,
} from 'react-native-responsive-screen';
import {useDispatch} from 'react-redux';

import Btn from '~/components/helpers/Button';
import {FEELING_MAPPED_KEY, feelings} from '~/constants/exercises';
import {setFilterBySelectedBadge} from '~/store/slices/exercises';
import {DEEP_BACKGROUND, FADED_BACKGROUND} from '~/styles/theme';

type BadgeInnerProps = {
  index: number;
  item: number;
  press?: () => void;
};

export default function BadgeInner({index, item, press}: BadgeInnerProps) {
  const styles = StyleSheet.create({
    badgeContainer: {
      marginTop: hp('2%'),
      marginBottom: hp('4%'),
    },
    badge: {
      backgroundColor: FADED_BACKGROUND[FEELING_MAPPED_KEY[item]],
      paddingHorizontal: 12,
      paddingVertical: 6,
      borderRadius: 20,
      marginRight: 10,
      borderColor: DEEP_BACKGROUND[FEELING_MAPPED_KEY[item]],
      borderWidth: 1,
    },
    badgeText: {
      color: DEEP_BACKGROUND[FEELING_MAPPED_KEY[item]],
      fontWeight: '800',
      fontSize: wp('4%'),
    },
  });

  const dispatch = useDispatch();
  const scale = useSharedValue<number>(1);

  const scaleStyles = useAnimatedStyle(() => ({
    transform: [{scale: scale.value}],
  }));

  function handleFilter(): void {
    setTimeout(() => {
      dispatch(setFilterBySelectedBadge(feelings[index]));
    }, 200);
  }

  return (
    <Btn
      style={styles.badgeContainer}
      onPress={() => {
        handleFilter();
        press && press();
      }}
      onPressIn={() => (scale.value = withSpring(1.05))}
      onPressOut={() => (scale.value = withSpring(1))}
    >
      <Animated.View
        style={[
          styles.badge,
          {marginLeft: index === 0 ? wp('5%') : 0},
          scaleStyles,
        ]}
      >
        <Text style={styles.badgeText}>{item}</Text>
      </Animated.View>
    </Btn>
  );
}
