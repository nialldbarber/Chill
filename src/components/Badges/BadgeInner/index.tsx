export {default as BadgeInner} from './BadgeInner';
