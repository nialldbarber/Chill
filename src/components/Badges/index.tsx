import React from 'react';

import BadgeInner from '~/components/Badges/BadgeInner';
import Scroll from '~/components/helpers/Scrollview';
import {feelings} from '~/constants/exercises';

export default function Badges({press}: {press: () => void}) {
  return (
    <Scroll h>
      {feelings.map((item, i) => (
        <BadgeInner key={i} {...{i, item, press}} />
      ))}
    </Scroll>
  );
}
