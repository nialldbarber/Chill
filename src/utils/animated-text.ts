export function getAnimatedTextFormatted(str: any): any {
  let formattedStr: any = str.value;
  formattedStr = str.replace(/NaN/g, '');
  return formattedStr;
}
