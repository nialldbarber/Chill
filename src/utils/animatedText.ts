import Animated from 'react-native-reanimated';

import {Instruct} from '~/screens/Exercise/Exercise';

type N = number;
type S = string;

export function getAnimatedTextFormatted(str: any): S {
  'worklet';
  let formattedStr = str?.value;
  formattedStr = formattedStr?.toString();
  formattedStr = str?.replace(/NaN/g, '');
  return formattedStr;
}

export const formatAnimatedStr = (
  str: S | N,
  instructions: Animated.SharedValue<Instruct>,
  type: N | undefined,
): S => {
  'worklet';

  let text: S | N = str;

  if (type === 1) {
    if (instructions.value < 1) {
      text = 'breathe in';
    } else if (instructions.value > 1 && instructions.value < 2) {
      text = 'hold';
    } else if (instructions.value > 2 && instructions.value < 3) {
      text = 'breathe out';
    } else if (instructions.value > 3 && instructions.value < 4) {
      text = 'hold';
    }
  } else if (type === 2) {
    if (instructions.value < 1) {
      text = 'breathe in';
    } else if (instructions.value > 1 && instructions.value < 2) {
      text = 'breathe out';
    } else if (instructions.value > 2 && instructions.value < 3) {
      text = 'hold';
    }
  } else if (type === 3) {
    if (instructions.value < 1) {
      text = 'breathe in';
    } else if (instructions.value > 1 && instructions.value < 2) {
      text = 'hold';
    } else if (instructions.value > 2 && instructions.value < 3) {
      text = 'breathe out';
    }
  } else if (type === 4) {
    if (instructions.value < 1) {
      text = 'breathe in';
    } else if (instructions.value > 1 && instructions.value < 2) {
      text = 'breathe out';
    }
  }

  return text.toString().replace(/NaN/g, '');
};
