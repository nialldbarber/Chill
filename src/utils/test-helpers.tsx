import React from 'react';

import {render} from '@testing-library/react-native';
import {Provider} from 'react-redux';

import {store} from '~/store/index';

export function renderWithWrapper(children) {
  return <Provider store={store}>{children}</Provider>;
}

export function matchSnapshot(component) {
  const shallow = render(renderWithWrapper(component)).toJSON();
  expect(shallow).toMatchSnapshot();
}
