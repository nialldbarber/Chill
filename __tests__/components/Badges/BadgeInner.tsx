import React from 'react';

import {render} from '@testing-library/react-native';

import {BadgeInner} from '~/components/Badges/BadgeInner';
import {matchSnapshot, renderWithWrapper} from '~/utils/test-helpers';

describe('<BadgeInner />', () => {
  it('should match snapshot', () => {
    matchSnapshot(<BadgeInner index={1} item={1} />);
  });

  it('should render 1 when provided', () => {
    const {getByText} = render(
      renderWithWrapper(<BadgeInner index={1} item={1} />),
    );
    const element = getByText('1');
    expect(element).toBeTruthy();
  });
});
