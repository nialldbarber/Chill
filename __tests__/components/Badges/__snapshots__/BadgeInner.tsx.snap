// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BadgeInner /> should match snapshot 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  hitSlop={
    Object {
      "bottom": undefined,
      "left": undefined,
      "right": undefined,
      "top": undefined,
    }
  }
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "marginBottom": 53.5,
      "marginTop": 26.5,
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": undefined,
          "borderColor": undefined,
          "borderRadius": 20,
          "borderWidth": 1,
          "marginRight": 10,
          "paddingHorizontal": 12,
          "paddingVertical": 6,
        },
        Object {
          "marginLeft": 0,
        },
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": undefined,
          "fontSize": 30,
          "fontWeight": "800",
        }
      }
    >
      1
    </Text>
  </View>
</View>
`;
